plugins/core/hu.bme.mit.gamma.expression.model/src/hu/bme/mit/gamma/expression/util/ExpressionTypeDeterminator2.java

```diff
@@ -0,0 +1,403 @@
+			}
+			return ExpressionModelDerivedFeatures.getTypeDefinition(type);
+		}
+		if (expression instanceof BooleanLiteralExpression) {
+			return factory.createBooleanTypeDefinition();
+		}
+		if (expression instanceof IntegerLiteralExpression) {
+			return factory.createIntegerTypeDefinition();
+		}
+		if (expression instanceof RationalLiteralExpression) {
+			return factory.createRationalTypeDefinition();
+		}
+		if (expression instanceof DecimalLiteralExpression) {
+			return factory.createDecimalTypeDefinition();
+		}
+		if (expression instanceof EnumerationLiteralExpression) {
+			return factory.createEnumerationTypeDefinition();
+		}
+		if (expression instanceof IntegerRangeLiteralExpression) {
+			return factory.createIntegerRangeTypeDefinition();
+		}
+		if (expression instanceof RecordLiteralExpression) {
+			return factory.createRecordTypeDefinition();
+		}
+		if (expression instanceof ArrayLiteralExpression) {
+			return factory.createArrayTypeDefinition();
+		}
+		if (expression instanceof DirectReferenceExpression) {
+			Type declarationType = ((DirectReferenceExpression) expression).getDeclaration().getType(); 
+			return ExpressionModelDerivedFeatures.getTypeDefinition(declarationType);
+		}
+		if (expression instanceof ElseExpression) {
+			return factory.createBooleanTypeDefinition();
+		}
+		if (expression instanceof BooleanExpression) {
+			Type declarationType = expressionUtil.getDeclaration(expression).getType();
+			return ExpressionModelDerivedFeatures.getTypeDefinition(declarationType);
+		}
+		if (expression instanceof PredicateExpression) {
+			return factory.createBooleanTypeDefinition();
+		}
+		if (expression instanceof QuantifierExpression) {
+			return factory.createBooleanTypeDefinition();
+		}
+		if (expression instanceof IfThenElseExpression) {
+			return factory.createBooleanTypeDefinition();
+		}
+		if (expression instanceof OpaqueExpression) {
+			// ?
+		}
+		if (expression instanceof UnaryPlusExpression) {
+			return getArithmeticUnaryType((UnaryPlusExpression) expression);
+		}
+		if (expression instanceof UnaryMinusExpression) {
+			return getArithmeticUnaryType((UnaryMinusExpression) expression);
+		}
+		if (expression instanceof SubtractExpression) {
+			return getArithmeticBinaryType((SubtractExpression) expression);
+		}
+		if (expression instanceof DivideExpression) {
+			return getArithmeticBinaryType((DivideExpression) expression);
+		}
+		if (expression instanceof ModExpression) {
+			return getArithmeticBinaryIntegerType((ModExpression) expression);
+		}
+		if (expression instanceof DivExpression) {
+			return getArithmeticBinaryIntegerType((DivExpression) expression);
+		}
+		if (expression instanceof AddExpression) {
+			return getArithmeticMultiaryType((AddExpression) expression);
+		}
+		if (expression instanceof MultiplyExpression) {
+			return getArithmeticMultiaryType((MultiplyExpression) expression);
+		}
+		if (expression instanceof ArrayAccessExpression) {
+			int depth = 0;
```

Comment:
Pretty complicated. Wouldn't checking the type of the accessed expression be enough? It must be of type array, and here we could return the `.elementType()` of the array type.